File: 2022/day05/day05.py
import re

stacks, moves = open("input.txt").read().split("\n\n")

sd = {k: [] for k in map(int, re.findall(r"\d+", stacks.split("\n")[-1]))}
for line in stacks.split("\n")[:-1]:
    for n, idx in enumerate(range(1, len(line), 4)):
        if line[idx] != " ":
            sd[n + 1].append(line[idx])

for move in moves.split("\n"):
    a, b, c = map(int, re.findall(r"\d+", move))

    # part 1
    sd[c] = [sd[b].pop(0) for _ in range(a)][::-1] + sd[c]

    # # part 2
    # sd[c] = [sd[b].pop(0) for _ in range(a)] + sd[c]
print("".join([v[0] for v in sd.values()]))

File: 2022/day05/day05.jl
function main()
    stacks, moves = filter(x -> !isempty(x), split(readchomp("input.txt"), "\n\n"))

    stacks = split(stacks, "\n")
    idxs = only.(findall(r"\d+", stacks[end]))
    stack_ary = hcat(map(x -> getindex.(x, idxs), stacks[1:end-1])...)
    ds = Dict(k => filter(x -> x != ' ', v)
              for (k, v) in zip(1:length(idxs), eachrow(stack_ary)))

    for move in split(moves, "\n")
        a, b, c = map(x -> parse(Int, x.match), eachmatch(r"\d+", move))

        prepend!(ds[c], reverse(splice!(ds[b], 1:a)))  # part 1
        # prepend!(ds[c], splice!(ds[b], 1:a))  # part 2
    end

    @show join(map(x -> ds[x][1], 1:length(idxs)))
end

main()
